src/adapter/src/catalog/migrate.rs

```diff
@@ -155,13 +156,482 @@ pub(crate) async fn migrate(
     //
     // Each migration should be a function that takes `tx` and `conn_cat` as
     // input and stages arbitrary transformations to the catalog on `tx`.
+    subsource_rewrite_v0_98(tx, &conn_cat, now)?;
+
     info!(
         "migration from catalog version {:?} complete",
         catalog_version
     );
     Ok(())
 }
 
+/// Inverts the dependency structure of subsources.
+///
+/// In previous versions of MZ, a primary source depended on its subsources.
+/// This migration inverts that relationship so that subsources depend on their
+/// primary source.
+///
+/// This function also adjusts items' `GlobalId`s such that the dependency graph
+/// can be inferred using the ordering of `GlobalId`s, i.e. all of an item's
+/// dependencies have `GlobalId`s less than its own.
+fn subsource_rewrite_v0_98(
+    tx: &mut Transaction<'_>,
+    conn_catalog: &ConnCatalog,
+    now: NowFn,
+) -> Result<(), anyhow::Error> {
+    use mz_sql::ast::UnresolvedItemName;
+    use mz_sql::catalog::CatalogItemType;
+    use mz_sql_parser::ast::RawItemName;
+    use mz_storage_types::connections::Connection;
+    use mz_storage_types::sources::GenericSourceConnection;
+
+    // A vector in the _new_ dependency order. Because the `vec` doesn't have
+    // any built-in de-duplication, we will need to deduplicate these values
+    // elsewhere.
+    let mut needs_new_id = vec![];
+
+    // The IDs of the sources whose subsources we are updating. We need to
+    // ensure we keep their IDs static.
+    let mut source_ids_of_updated_subsources = BTreeSet::new();
+
+    // Collect all items that we need to update in the txn; do this en masse
+    // because individually updating items can perform poorly.
+    let mut updated_items = BTreeMap::new();
+
+    // Get all sources in this TXN.
+    let mut source_map: BTreeMap<_, _> = tx
+        .get_items()
+        .into_iter()
+        .filter(|item| conn_catalog.get_item(&item.id).item_type() == CatalogItemType::Source)
+        .map(|item| (item.id, item))
+        .collect();
+
+    // Iterate over all sources in the catalog.
```

Comment:
What triggers sources to be already in the Transaction? Down below you have a
```source_map
    .remove(&primary_source_id)
    .expect("source exists");
```
which makes it seem that all sources are part of the transaction? Why do we need to do this list/filter operation to find sources on both the txn and the catalog?